packages/plugin-dom/DomRenderer.ts

```diff
@@ -40,68 +29,61 @@ export class DomRenderer extends Renderer<
         }
 
         // Don't render `root` itself if already rendered, render its children.
-        const renderedRoot = VDocumentMap.toDom(root);
-        const parentNode = renderedRoot ? renderedRoot : target;
-        const firstVNode = renderedRoot ? root.firstChild() : root;
-
-        if (!firstVNode) return;
+        const domRoot = VDocumentMap.toDom(root);
+        const domParent = domRoot ? domRoot : target;
+        const firstNode = domRoot ? root.firstChild() : root;
 
-        const rootContext: DomRenderingContext = {
-            root: root,
-            currentNode: firstVNode,
-            parentNode: parentNode,
-        };
-        this._contextStack.push(rootContext);
+        if (!firstNode) return;
 
-        let currentContext: DomRenderingContext = { ...rootContext };
-        do {
-            currentContext = this.renderNode(currentContext);
-        } while ((currentContext = this._nextRenderingContext(currentContext)));
+        let node = firstNode;
+        let rendered = [];
+        while (node) {
+            if (!rendered.includes(node)) {
+                rendered = [...rendered, ...this.renderNode(node, domParent).keys()];
```

Comment:
I would explicitely call `renderNode` in its own statement, it is in fact the most important part of the algorithm.